lib/awful/keygrabber.lua

```diff
@@ -45,7 +207,411 @@ function keygrabber.stop(g)
+--
+-- @property keybindings
+-- @param table
+-- @see export_keybindings
+-- @see root_keybindings
+
+--- Define a list of keys that are allowed instead of using `release_key`.
+--
+-- When defined, this property allows to define an implicit way to release the
+-- keygrabber. It helps save some boilerplate code in the handler callbacks.
+--
+-- @DOC_text_awful_keygrabber_allowed_keys_EXAMPLE@
+--
+-- @property allowed_keys
+-- @tparam[opt=nil] table|nil allowed_keys The list of keys.
+
+--- The sequence recorded from the current acquisition.
+--
+-- In this example, the `stop_callback` is used to retrieve the final key
+-- sequence.
+--
+-- Please note that backspace will remove elements from the sequence and
+-- named keys and modifiers are ignored.
+--
+-- @DOC_text_awful_keygrabber_autostart_EXAMPLE@
+--
+-- @property sequence
+-- @param string
+--
+
+--- The current (running) instance of this keygrabber.
+--
+-- The keygrabber instance is created when the keygrabber starts. It is an object
+-- mirroring this keygrabber object, but where extra properties can be added. It
+-- is useful to keep some contextual data part of the current transaction without
+-- poluting the original object of having extra boilerplate code.
+--
+-- @tfield keygrabber current_instance
+-- @see property::current_instance
+
+--- The global signal used to track the `current_instance`.
+--
+-- @signal property::current_instance
+
+--- If a keygrabber is currently running.
+-- @tfield boolean is_running
+
+--- Start the keygrabber.
+--
+-- Note that only a single keygrabber can be started at any one time. If another
+-- keygrabber (or this one) is currently running. This method returns false.
+--
+-- @treturn boolean If the keygrabber was successfully started.
+function keygrabber:start()
+    if self.grabber or keygrab.current_instance then
+        return false
+    end
+
+    self.current_instance = setmetatable({}, {
+        __index = self,
+        __newindex = function(tab, key, value)
+            if keygrabber["set_"..key] then
+                self[key](self, value)
+            else
+                rawset(tab, key, value)
+            end
+        end
+    })
+
+    self.sequence = ""
+
+    if self.start_callback then
+        self.start_callback(self.current_instance)
+    end
+
+    self.grabber = keygrab.run(function(...) return runner(self, ...) end)
+
+    -- Ease making keygrabber that wont hang forever if no action is taken.
+    if self.timeout and not self._private.timer then
+        self._private.timer = gtimer {
+            timeout     = self.timeout,
+            single_shot = true,
+            callback    = function()
+                if self.timeout_callback then
+                    self.timeout_callback(self)
+                end
+                self:stop()
+            end
+        }
+    end
+
+    if self._private.timer then
+        self._private.timer:start()
+        assert(self._private.timer.data.source_id)
```

Comment:
Why this `assert`? Are you testing that `:start()` works correctly?!?
Why not use `.is_running` instead?

(Slightly unrelated: Should `.data` be renamed to `._private`?)